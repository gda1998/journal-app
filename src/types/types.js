const types = {
    login: '[Auth] Login',
    logout: '[Auth] Logout',

    uiSetError: '[UI] SetError',
    uiRemoveError: '[UI] RemoveError',
    uiStartLoading: '[UI] StartLoading',
    uiFinishLoading: '[UI] FinishLoading',

    notesAddNew: '[Notes] New note',
    notesActive: '[Notes] Set active note',
    notesLoad: '[Notes] Load notes',
    notesUpdate: '[Notes] Update note',
    notesFileUrl: '[Notes] Update note image url',
    notesDelete: '[Notes] Delete note',
    notesLogoutCleaning: '[Notes] Logout Cleaning'
};

export default types;