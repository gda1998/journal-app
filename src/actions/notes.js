import { db } from '../firebase/firebaseConfig';
import types from '../types/types';
import {
    addDoc,
    collection,
    doc,
    updateDoc,
    deleteDoc,
    // getDoc
} from 'firebase/firestore';
import Swal from 'sweetalert2';
import { swalAlert, swalLoading } from '../helpers/swalToast';
import loadNotes from '../helpers/loadNotes';
import fileUpload from '../helpers/fileUpload';
// import fileDelete from '../helpers/fileDelete';

export const startNewNoteAction = () => {
    return (dispatch, getState) => {
        const { uid } = getState().auth;
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        };

        const notesRef = collection(db, `${uid}/journal/notes`);
        return addDoc(notesRef, newNote)
            .then(docRef => {
                dispatch(activeNoteAction(docRef.id, newNote));
                dispatch(addNewNoteAction(docRef.id, newNote));
            })
            .catch(console.error);
    }
}

export const activeNoteAction = (id, note) => ({
    type: types.notesActive,
    payload: { id, ...note }
});

export const addNewNoteAction = (id, note) => ({
    type: types.notesAddNew,
    payload: { id, ...note }
});

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotesAction(notes));
    }
}

export const setNotesAction = (notes) => ({
    type: types.notesLoad,
    payload: notes
});

export const startUpdateNote = (note) => {
    return (dispatch, getState) => {
        const { uid } = getState().auth;
        !note.url && (note.url = null);
        const noteRef = doc(db, `${uid}/journal/notes/`, note.id);
        updateDoc(noteRef, note);
        dispatch(refreshNoteAction(note));
        swalAlert('Updated!', 'The note was updated', 'success');
    }
}

export const refreshNoteAction = (note) => ({
    type: types.notesUpdate,
    payload: note
});

export const startUploadingImgAction = (file) => {
    return async (dispatch, getState) => {
        const { active: activeNote } = getState().notes;
        swalLoading('Uploading', 'Please wait...');

        const fileUrl = await fileUpload(file);
        activeNote.url = fileUrl;

        Swal.close();
        await dispatch(startUpdateNote(activeNote));
    }
}

export const startDeleteNote = (id) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;
        try {
            const ref = await doc(db, uid, `journal/notes/${id}`);
            // const noteRef = await getDoc(ref);

            // Delete the image from cloudinary and from the db
            // fileDelete(noteRef.data().url)
            await deleteDoc(ref);

            dispatch(deleteNote(id));
        } catch (error) {
            console.error('Error al eliminar la nota: ', error);
        }
    }
}

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
});

export const notesLogout = () => ({
    type: types.notesLogoutCleaning
})