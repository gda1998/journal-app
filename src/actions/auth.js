import { 
    createUserWithEmailAndPassword,
    getAuth,
    signInWithPopup,
    signInWithEmailAndPassword,
    signOut,
    updateProfile 
} from 'firebase/auth';
import { googleAuthProvider } from '../firebase/firebaseConfig';
import { finishLoadingAction, startLoadingAction } from './ui';
import { swalAlert } from '../helpers/swalToast';
import types from '../types/types';
import { notesLogout } from './notes';

export const startRegisterAction = ({ email, password, name }) => {
    return (dispatch) => {
        const auth = getAuth();
        dispatch( startLoadingAction() );
        createUserWithEmailAndPassword(auth, email, password)
        .then( async({ user }) => {
            await updateProfile(user, { displayName: name });
            dispatch( login(user.uid, user.displayName) );
        })
        .catch( e => swalAlert('Error!', e, 'error') )
        .finally( () => dispatch( finishLoadingAction() ) );
    };
};

export const startLoginEmailPassword = ({ email, password }) => {
    return (dispatch) => {
        const auth = getAuth();
        dispatch( startLoadingAction() );
        return signInWithEmailAndPassword(auth, email, password)
        .then( ({user}) => dispatch( login(user.uid, user.displayName) ) )
        .catch( e => swalAlert('Error!', e, 'error') )
        .finally( () => dispatch( finishLoadingAction() ) );
    };
};

export const startGoogleLogin = () => {
    return (dispatch) => {
        const auth = getAuth();
        signInWithPopup(auth, googleAuthProvider)
        .then( ({user}) => {
            dispatch( login(user.uid, user.displayName) )
        });
    }
}

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
});

export const logout = () => ({
    type: types.logout
})

export const startLogout = () => {
    return async(dispatch) => {
        await signOut( getAuth() );
        dispatch( logout() );
        dispatch(  notesLogout() );
    }
}