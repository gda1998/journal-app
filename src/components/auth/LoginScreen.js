import React from 'react';
import { useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import { login } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';

export const LoginScreen = () => {

    const dispatch = useDispatch();

    const [ formValues, handleInputChange ] = useForm({
        email: 'gabriel@gmail.com',
        password: '123456'
    });

    const { email, password } = formValues;

    const handleLogin = e => {
        e.preventDefault();
        dispatch( login(1234, 'Hernando') );   
    }

    return (
        <>
            <h3 className="auth__title">Login</h3>
            <form onSubmit={ handleLogin }>

                <input
                    autoComplete="off"
                    className="auth__input"
                    type="email"
                    placeholder="Email"
                    name="email"
                    value={ email }
                    onChange={ handleInputChange }
                />
                <input
                    className="auth__input"
                    type="password"
                    placeholder="Password"
                    name="password"
                    value={ password }
                    onChange={ handleInputChange }
                />
                <button type="submit" className="btn btn-primary btn-block">
                    Login
                </button>

                <div className="auth__social-networks">
                    <p>Login with social networks</p>
                    <div className="google-btn">
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>{/* /.google-btn */}
                </div>{/* /.auth__social-networks */}

                <Link to="/auth/register" className="link">
                    Create a new account
                </Link>

            </form>
        </>
    );
}
