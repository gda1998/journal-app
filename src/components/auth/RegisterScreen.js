import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import validator from 'validator';
import { useForm } from '../../hooks/useForm';
import { startRegisterAction } from '../../actions/auth';
import { removeErrorAction, setErrorAction } from '../../actions/ui';

export const RegisterScreen = () => {
    const dispatch = useDispatch();
    const { msgError, loading } = useSelector( state => state.ui );
    const [ formValues, handleInputChange ] = useForm({
        name: '',
        email: '',
        password: '',
        password2: ''
    });
    const { name, email, password, password2 } = formValues;

    const handleRegister = (e) => {
        e.preventDefault();
        isFormValid() && dispatch( startRegisterAction(formValues) );
    }

    const isFormValid = () => {
        if( !validator.isEmail(email) ){
            dispatch( setErrorAction('Email Invalid') );
            return false;
        }
        if( password !== password2 || password.length < 6 ){
            dispatch( setErrorAction('Password should be at least 6 characters and match each other') );
            return false;
        }
        dispatch( removeErrorAction() );
        return true;
    }

    return (
        <>
            <h3 className="auth__title">Register</h3>
            <form
                className="animate__animated animate__fadeIn animate__faster"
                onSubmit={ handleRegister }
            >
                {
                    msgError && 
                    <div className="auth__alert-error">
                        { msgError }
                    </div>
                }

                <input
                    autoComplete="off"
                    className="auth__input"
                    maxLength={255}
                    name="name"
                    onChange={ handleInputChange }
                    pattern="[a-zA-ZÀ-ÿ\u00f1\u00d1]+(\s*[a-zA-ZÀ-ÿ\u00f1\u00d1]*)*"
                    placeholder="Name"
                    required
                    title="El campo debe tener solamente caracteres alfabéticos."
                    type="text"
                    value={ name }
                />

                <input
                    autoComplete="off"
                    className="auth__input"
                    maxLength={255}
                    name="email"
                    onChange={ handleInputChange }
                    pattern="[a-zA-Z0-9_]+([.\-][a-zA-Z0-9_]+)*@[a-zA-Z0-9_]+([.][a-zA-Z0-9_]+)*[.][a-zA-Z]{1,5}"
                    placeholder="Email"
                    required
                    title="ej. example@email.com"
                    type="email"
                    value={ email }
                />
                <input
                    className="auth__input"
                    minLength={ 6 }
                    maxLength={255}
                    name="password"
                    onChange={ handleInputChange }
                    placeholder="Password"
                    required
                    type="password"
                    value={ password }
                />
                <input
                    className="auth__input"
                    minLength={ 6 }
                    maxLength={255}
                    name="password2"
                    onChange={ handleInputChange }
                    placeholder="Confirm"
                    required
                    type="password"
                    value={ password2 }
                />
                <button 
                    className="btn btn-primary btn-block mb-5"
                    disabled={ loading }
                    type="submit" 
                >
                    Register
                </button>

                <Link to="/auth/login" className="link">
                    Already registered?
                </Link>

            </form>
        </>
    );
}