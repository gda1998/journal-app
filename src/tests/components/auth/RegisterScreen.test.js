import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { startRegisterAction } from '../../../actions/auth';
import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import types from '../../../types/types';

jest.mock('../../../actions/auth', () => ({
    ...jest.requireActual('../../../actions/auth'),
    startRegisterAction: jest.fn()
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const store = mockStore({
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
});
// store.dispatch = jest.fn(); // * Solo se usa cuando las acciones son asincronas


describe('Pruebas en <RegisterScreen />', () => {
    let wrapper;
    const getWrapper = (store) => mount(
        <Provider store={store}>
            <MemoryRouter>
                <RegisterScreen />
            </MemoryRouter>
        </Provider>);

    const targetObject = (name, value) => ({
        target: { name, value }
    });

    beforeEach( () => {
        wrapper = getWrapper(store);
        jest.clearAllMocks();
        store.clearActions();
    });

    test('Debe de mostrarse correctamente', () => { 
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de hacer el dispatch con setErrorAction si el email es invalido', () => { 
        wrapper.find('input').at(1).simulate('change', targetObject('email', 'abcd'));
        wrapper.find('form').simulate('submit', { preventDefault(){} });
        
        const actions = store.getActions();

        expect(actions[0]).toMatchObject({
            type: types.uiSetError,
            payload: 'Email Invalid'
        });
    });

    test('Debe de mostrar un error si el email no esta validado', () => { 
        const store = mockStore({
            auth: {},
            ui: {
                loading: false,
                msgError: 'Email Invalid'
            }
        });

        const wrapper = getWrapper(store);
        const errorDiv = wrapper.find('.auth__alert-error');
        expect(errorDiv.exists()).toBe(true);
        expect(errorDiv.text()).toBe('Email Invalid');
    });

    test('Debe de funcionar el handleInputChange del formulario', () => { 
        wrapper.find('input[name="name"]').simulate('change', targetObject('name', 'Gabriel Diaz'));
        expect(wrapper.find('input[name="name"]').prop('value')).toBe('Gabriel Diaz');
    });

    test('Debe de llamar al dispatch startRegisterAction si el formulario esta bien llenado', () => { 
        const store = mockStore({
            auth: {},
            ui: {
                loading: false,
                msgError: null
            }
        });
        store.dispatch = jest.fn();
        const wrapper = getWrapper(store);

        wrapper.find('input[name="name"]').simulate('change', targetObject('name', 'Gabriel Diaz'));
        wrapper.find('input[name="email"]').simulate('change', targetObject('email', 'gabriel@gmail.com'));
        wrapper.find('input[name="password"]').simulate('change', targetObject('password', '123456'));
        wrapper.find('input[name="password2"]').simulate('change', targetObject('password2', '123456'));

        wrapper.find('form').simulate('submit', { preventDefault(){} });

        expect(startRegisterAction).toHaveBeenCalled();
    });
});