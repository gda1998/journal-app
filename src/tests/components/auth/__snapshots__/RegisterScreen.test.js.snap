// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <RegisterScreen /> Debe de mostrarse correctamente 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      maxLength={255}
      name="name"
      onChange={[Function]}
      pattern="[a-zA-ZÀ-ÿ\\\\u00f1\\\\u00d1]+(\\\\s*[a-zA-ZÀ-ÿ\\\\u00f1\\\\u00d1]*)*"
      placeholder="Name"
      required={true}
      title="El campo debe tener solamente caracteres alfabéticos."
      type="text"
      value=""
    />
    <input
      autoComplete="off"
      className="auth__input"
      maxLength={255}
      name="email"
      onChange={[Function]}
      pattern="[a-zA-Z0-9_]+([.\\\\-][a-zA-Z0-9_]+)*@[a-zA-Z0-9_]+([.][a-zA-Z0-9_]+)*[.][a-zA-Z]{1,5}"
      placeholder="Email"
      required={true}
      title="ej. example@email.com"
      type="email"
      value=""
    />
    <input
      className="auth__input"
      maxLength={255}
      minLength={6}
      name="password"
      onChange={[Function]}
      placeholder="Password"
      required={true}
      type="password"
      value=""
    />
    <input
      className="auth__input"
      maxLength={255}
      minLength={6}
      name="password2"
      onChange={[Function]}
      placeholder="Confirm"
      required={true}
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block mb-5"
      disabled={false}
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;
